File: Common/Serialization/ObjectTraverseProvider.vb
Namespace Common

    Public MustInherit Class ObjectTraverseProvider

        Protected MustOverride Function GetSetPropertiesOverride(ByVal Handler As PropertyTraverseHandler) As IEnumerable(Of Void)

    End Class


    Public Class PropertyTraverseHandler



    End Class

    Public Structure PropertyTraverseCurrent

        Public Sub New(ByVal Handler As PropertyTraverseHandler)
            Me._Handler = Handler
        End Sub

#Region "Handler Property"
        Private ReadOnly _Handler As PropertyTraverseHandler

        Public ReadOnly Property Handler As PropertyTraverseHandler
            Get
                Return Me._Handler
            End Get
        End Property
#End Region

    End Structure

    Public Interface PropertyTraverseProcessor

        Sub Process(Of T)(ByVal Prop As PropertyTraverseHandler)

    End Interface

End Namespace

File: Common/Serialization/ObjectTraverseProvider.cs
using System.Collections.Generic;

namespace Ks.Common
{
    public abstract class ObjectTraverseProvider
    {
        protected abstract IEnumerable<Void> GetSetPropertiesOverride(PropertyTraverseHandler Handler);
    }


    public class PropertyTraverseHandler
    {
    }

    public struct PropertyTraverseCurrent
    {
        public PropertyTraverseCurrent(PropertyTraverseHandler Handler)
        {
            this._Handler = Handler;
        }

        private readonly PropertyTraverseHandler _Handler;

        public PropertyTraverseHandler Handler
        {
            get
            {
                return this._Handler;
            }
        }
    }

    public interface PropertyTraverseProcessor
    {
        void Process<T>(PropertyTraverseHandler Prop);
    }
}
